### ISSUE TITLE
Cannot debug Electron app on Windows

### ISSUE THREAD

**Issue by JimiC** [Microsoft/vscode#7583]
- VSCode Version: 1.2.0 (also latest 1.3.0-insiders)
- OS Version: Windows 10 (build 10586.318)

Steps to Reproduce:

Using the following `launch.json` configuration:

```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "node",
            "request": "launch",
            "program": "${workspaceRoot}/src/scripts/main.ts",
            "stopOnEntry": false,
            "args": [],
            "cwd": "${workspaceRoot}/src",
            "preLaunchTask": "compile",
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron",
            "runtimeArgs": [
                "--enable-logging"
            ],
            "env": {},
            "externalConsole": false,
            "sourceMaps": true,
            // win32 & linux
            "outDir": "${workspaceRoot}/src/bin/resources/app"
            // darwin
            //"outDir": "${workspaceRoot}/src/bin/Electron.app/Contents/Resources/app"
        },
        {
            "name": "Attach",
            "type": "node",
            "request": "attach",
            "port": 5858,
            "address": "localhost",
            "restart": false,
            "sourceMaps": false,
            "outDir": null,
            "localRoot": "${workspaceRoot}",
            "remoteRoot": null
        }
    ]
}
```

I'm getting the following error message 
`Cannot launch debug target (spawn ${workspaceRoot}/node_modules/.bin/electron ENOENT)`
when trying to debug, either via F5 or directly from Debugger window.

If I set `"externalConsole": true` then the Debugger starts and opens an external console.
This behavior is not noticeable on Linux or OS X (setting `externalConsole` either to false or true works as intended).

**Comment by weinand** [comment]
@JimiC the error 'ENOENT' means that there is no executable `${workspaceRoot}/node_modules/.bin/electron`. Did you verify that an executable exists under that path?

**Comment by JimiC** [comment]
Yeap. Surprisingly when I switch to `"externalConsole": true` it works and as I mentions on Linux and OS X I have no issues with either `"externalConsole": true` or `"externalConsole": false`. This is specific to Windows and with `"externalConsole": false`.

**Comment by JimiC** [comment]
@weinand If I use `"runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd"` it starts but still opens an external console when it shouldn't.

**Comment by weinand** [comment]
@JimiC if an external console is used on WIndows, the command interpreter running in that console is probably automatically adding a '.exe' or '.cmd' to the path specified whereas the internal launching mechanism doesn't do this.

From the two scripts (electron.cmd and electron.sh) I get the impression that you could launch electron directly without specifying a `runtimeExecutable` because both scripts just rely on node.js and a 'cli.js'.
Did you try to specify  `${workspaceRoot}/electron-prebuilt/cli.js` for `program` and add your `${workspaceRoot}/src/scripts/main.ts` to the `args` array?

**Comment by JimiC** [comment]
@weinand First the correct value for `program` is `"${workspaceRoot}/node_modules/electron-prebuilt/cli.js"` (no offence).
Now in that case I have to comment out the `runtimeExecutable` and `runtimeArgs`. After doing that it starts but immediately throws an exception `A Javascript error occurred in the min process` (Syntax Error for `import`) because it tries to debug a `.js` and I have specified a `.ts` (Note: I use ES6 syntax in Typescript).

**Comment by weinand** [comment]
@JimiC yes, you do not need to specify the runtimeExecutable because you need 'node' in this case which is the default anyway.
You can move your 'runtimeArgs' to 'args'. 
Yes, Node debug cannot automatically map the argument 'main.ts' to the correct 'main.js'. It can only do this for the 'program' attribute.

BTW, VS Code node debug builds the final launch command from the attributes in the following way:

```
<runtimeExecutable> <runtimeArgs> --debug-brk <program> <args>
```

**Comment by JimiC** [comment]
@weinand That's good to know but the fact here is to be able to debug using 'electron' instead of 'node'. What you said about console automatically adding '.exe' is true.

Because atm I'm afk I'll come back later with some screenies. At the meantime feel free to check out the sample project.

**Comment by JimiC** [comment]
@weinand 

Pic 1
```
"runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron"
"externalConsole": false
```
![untitled1](https://cloud.githubusercontent.com/assets/1716442/16018247/50bc04ec-31ac-11e6-9902-a6c831ced139.png)

Pic 2
```
"runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron"
"externalConsole": true
```
![untitled2](https://cloud.githubusercontent.com/assets/1716442/16018248/50bde6d6-31ac-11e6-9f23-97f24381046c.png)

Pic 3
```
"runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd"
"externalConsole": false
```
![untitled3](https://cloud.githubusercontent.com/assets/1716442/16018249/50db3272-31ac-11e6-9d9d-30cab60acaba.png)

Pic 4
```
"runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd"
"externalConsole": true
```
![untitled4](https://cloud.githubusercontent.com/assets/1716442/16018246/50b83722-31ac-11e6-868a-a3c54dd4265e.png)

Something worth mentioning is that when using settings of Pic 3 the console is launched as an Electron process (weird?!)

Hoping the screenies will give an insight to the strange (bug?) behavior.

**Comment by JimiC** [comment]
Just FYI is I use
```
"program": "${workspaceRoot}/node_modules/electron-prebuilt/cli.js",
 "args": ["bin/resources/app/main.js", "--enable-logging"],
```
I get the same behavior as Pic 3.

**Comment by JimiC** [comment]
So I modified my `launch.json` configuration to look like this:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch",
            "type": "node",
            "request": "launch",
            "program": "${workspaceRoot}/src/evetron/scripts/main.ts",
            "stopOnEntry": false,
            "args": [],
            "cwd": "${workspaceRoot}",
            "preLaunchTask": null,
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron",
            "runtimeArgs": [
                "--enable-logging"
            ],
            "env": {},
            "externalConsole": false,
            "sourceMaps": true,
            "outDir": "${workspaceRoot}/src/evetron/bin/resources/app",
            "windows": {
                "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd",
                "runtimeArgs": []
            },
            "osx": {
                "outDir": "${workspaceRoot}/src/evetron/bin/Electron.app/Contents/Resources/app"
            }
        },
        {
            "name": "Attach",
            "type": "node",
            "request": "attach",
            "port": 5858,
            "address": "localhost",
            "restart": false,
            "sourceMaps": false,
            "outDir": null,
            "localRoot": "${workspaceRoot}",
            "remoteRoot": null
        }
    ]
}
```
It partially does what I'm looking for except that on Windows I don't have `logging` in the Debug Console.

**Comment by weinand** [comment]
@JimiC thanks a lot for this detailed analysis. I still have to understand whether this is a VS Code bug, an Electron issue, or some platform limitation.

But you seem to be unblocked, right?

And thanks for trying the new OS specific section.

**Comment by JimiC** [comment]
@weinand I'm not sure either if it's a
> VS Code bug, an Electron issue, or some platform limitation.

What troubles me is that if I open a Console on Windows and execute `${workspaceRoot}/node_modules/.bin/electron` it goes and executes as if I had executed `${workspaceRoot}/node_modules/.bin/electron.cmd`.

This has to be a VS Code misbehavior.

**Comment by weinand** [comment]
@JimiC but isn't this the normal command prompt behaviour of Windows?
If you have a file 'foo.cmd' and type 'foo' in the command prompt, 'foo.cmd' is executed.

**Comment by JimiC** [comment]
@weinand Indeed. But I suspect that Code doesn't executes the 'foo.cmd' on Windows but tries to execute 'foo' which is for `Unix` like OS. You have to tell it explicitly the `runtimeExecutable` file extension. 
If I were in you (and I'm not) I would have add a file extension check for Windows and add the `.cmd` if the user hasn't specified it. It's an internal handling (which you may not prefer) but improves the user experience with Code and Electron.

**Comment by weinand** [comment]
@JimiC in addition to '.cmd' I would have to check for '.bat' and '.exe' and may be even more. And I would have to implement the same order for this sequence as the windows command prompt...

**Comment by JimiC** [comment]
@weinand I totally understand that there is more to it than meets the eye. OS properties solves that for now.

**Comment by sukhanna** [comment]
I face similar issue on windows 7. I tried "console" : "integratedTerminal" and it works fine.

**Comment by weinand** [comment]
@georgiosd please add a `"trace": "all"`option to your launch config and share the trace output (if possible) here. Thanks.

**Comment by georgiosd** [comment]
@weinand can you please point me to the trace output path?

**Comment by weinand** [comment]
the trace output shows up in the debug console.

**Comment by georgiosd** [comment]
Hm, all I get is `22352: disconnectRequest: send response` in the debug console

**Comment by georgiosd** [comment]
Got it!
```
10932: _registerSourceMap: c:/users/xxxx/electron/dist/main.ts -> c:/users/xxx/electron/dist/main.js
```

**Comment by weinand** [comment]
Before getting the source maps and TS to work, I suggest to start with plain JavaScript.
Please set sourceMaps to false and and set `program`to the JavaScript path. 
Does this work?

In http://electron.rocks/debugging-electron-in-vs-code/ I noticed that they pass a "." as the first argument to electron...?

**Comment by georgiosd** [comment]
Yes, this works:

```
{
            "name": "Debug main process",
            "type": "node",
            "request": "launch",
            "program": "${workspaceRoot}/dist/main.js",
            "stopOnEntry": false,
            "args": [],
            "cwd": "${workspaceRoot}",
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd",
            "runtimeArgs": [
                "--enable-logging"
            ],
            "env": {},
            "sourceMaps": false,
            "outFiles": [
                "${workspaceRoot}/dist/**/*.js"
            ],
            "internalConsoleOptions": "openOnSessionStart",
            "console": "integratedTerminal",
            "trace": "all"
        },
```

**Comment by JimiC** [comment]
@georgiosd This probably means that you are not generation the source maps correctly. Find the souce map to `main.ts` and make sure the `source` points to the correct path. This is a tricky subject as I banged my head for hours to figure it out.

**Comment by georgiosd** [comment]
Well, isn't that what the `outFiles` property is supposed to do? I banged my head for hours to get the sourcemaps to work with Chrome debugger :)

**Comment by JimiC** [comment]
Have a look at my [`gulpfile.js`](https://github.com/JimiC/evetron/blob/master/gulpfile.js). Maybe it will give  you a hint.

**Comment by weinand** [comment]
"outFiles" tries to help for the (reverse) TS -> JS mapping because there is nothing in the TS source that points to the generated JavaScript. With "outFiles" the node debugger can at least search in all JS files under the glob patterns for references to the originating source. But this still requires that the paths in the source maps correctly point to the source.

How are you generating the source maps?

**Comment by georgiosd** [comment]
@JimiC thanks, I'll have a look tomorrow with clear head

@weinand This is my gulp task:

```
gulp.task('electron:transpile:ts', () => {
    var ts = require('gulp-typescript');
	var project = ts.createProject('./tsconfig.json');
	var tsResult = project.src()
        .pipe(sourcemaps.init())
		.pipe(project());

    return tsResult.js
        .pipe(sourcemaps.write('.', { 
            //sourceRoot: '/',
            mapSources: (sourcePath) => sourcePath.replace('../src/', '')
        }))
		.pipe(gulp.dest('./dist'));
});
```

nothing special in tsconfig.json but here it is:

```
{
  "compilerOptions": {
    "target": "es5",
    "rootDir": "src",
    "outDir": "dist",
    "moduleResolution": "node",
    "module": "commonjs",
    "declaration": false,
    "removeComments": false,
    "noLib": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "pretty": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitUseStrict": false,
    "noFallthroughCasesInSwitch": true
  },
  "exclude": [
    "bower_components",
    "node_modules",
    "dist",
    "typings/main",
    "typings/main.d.ts",
    "dist/assets/js/vendor",
    "packages"
  ]
}
```
Map file for `main.ts` is:

```
{"version":3,"sources":["main.ts"],"names":[],"mappings":"...","file":"main.js","sourcesContent":[null]}
```

**Comment by weinand** [comment]
The gulpfile.js looks ok, but since the mapping doesn't work, there seems to be a problem.

Please verify the following in the main.js source map:
If you concatenate the sourceRoot and the entries in the sources array, you get a relative path.
Append this path to the location (folder) of the source map. This should result in the absolute path of main.ts.

**Comment by JimiC** [comment]
@weinand @georgiosd As I said the issue is in the source map `sourceRoot` path. I had this issue myself.

**Comment by JimiC** [comment]
@georgiosd The `sourcePath` should not be empty put point to the scripts folder.

**Comment by weinand** [comment]
@JimiC sure, and to find that issue I asked @georgiosd to verify the resulting path of the sourcemap.
The tricky thing about relative paths is that the path is relative to the location of the sourcemap itself. So it cannot be constant but varies with the location of the sourcemap.

**Comment by weinand** [comment]
One solution is to use something like this:
```ts
.pipe(sourcemaps.write({
      sourceRoot: function (file) {
        return path.relative(file.relative, path.join(file.cwd, 'src'));
      }
    }))
```

In other cases it helps just to use a relative path for the sourceRoot itself:
```ts
.pipe(sourcemaps.write('.', {
      includeContent: false,
      sourceRoot: './'
    }))
```

**Comment by georgiosd** [comment]
@JimiC I see you don't actually have a launch configuration for debugging the renderer processes :)

@weinand thank you for the insights, the solution deemed less complex than I originally thought and this seems to work for debugging both the main and renderer processes:

gulp task:
```
.pipe(sourcemaps.write('.', {
            sourceRoot: './'
        }))
```

launch.json:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug main process",
            "type": "node",
            "request": "launch",
            "program": "${workspaceRoot}/src/main.ts",
            "stopOnEntry": false,
            "args": [],
            "cwd": "${workspaceRoot}/dist",
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd",
            "runtimeArgs": [
                "--enable-logging"
            ],
            "env": {},
            "sourceMaps": true,
            "outFiles": [
                "${workspaceRoot}/dist/**/*.js"
            ],
            "internalConsoleOptions": "openOnSessionStart",
            "console": "integratedTerminal",
            "preLaunchTask": "build"
        },
        {
            "name": "Debug renderer process",
            "type": "chrome",
            "request": "launch",
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd",
            "runtimeArgs": [
                "${workspaceRoot}/dist",
                "--enable-logging",
                "--remote-debugging-port=9222"
            ],
            "webRoot": "${workspaceRoot}/dist",
            "sourceMaps": true,
            "internalConsoleOptions": "openOnSessionStart"
        }
    ]
}
```
I found that all paths in the chrome task must point to "dist" folder otherwise I get a node-sass error that it doesn't support my platform :)

Thank you fellas!

**Comment by JimiC** [comment]
@georgiosd Haven't got to that part yet, but glad you figured it out. 😄

**Comment by georgiosd** [comment]
@JimiC oh I'm glad too, I spent at least a day on this lol :)
Anyway, the launch config is above if you wanna try it.

**Comment by georgiosd** [comment]
@weinand are you feeling brave in the new year? (happy new year)

I've changed my build process from gulp to webpack so I can have all that HMR goodness but I can't get the sourcemaps to work for the `node` task (haven't tried the chrome one yet). As you probably are aware, webpack bundles all the files into one and does produce a sourcemap but it has `webpack://` urls, in which case we'd need something like `sourceMapPathOverrides` but the IDE tells me it's not supported for this type of task.

Any ideas?

This is my node config:

```
 {
            "name": "Debug main process",
            "type": "node",
            "request": "launch",
            "program": "${workspaceRoot}/src-main/main.ts",
            "stopOnEntry": false,
            "args": [],
            "cwd": "${workspaceRoot}/dist",
            "runtimeExecutable": "${workspaceRoot}/node_modules/.bin/electron.cmd",
            "runtimeArgs": [
                "--enable-logging"
            ],
            "env": {},
            "sourceMaps": true,
            "outFiles": [
                "${workspaceRoot}/dist/main.electron.js"
            ],
            "internalConsoleOptions": "openOnSessionStart",
            "console": "externalTerminal"
        },
```
Using `trace:all` gives:

```
25176: _findSourceMapUrl: source map url found at end of generated file 'c:/dev/electron/dist/main.electron.js'
25176: _registerSourceMap: c:/dev/electron/dist/webpack:/webpack/bootstrap 10128bc400a0060e0bdc -> c:/dev/electron/dist/main.electron.js
25176: _registerSourceMap: webpack:///(webpack)/hot/log-apply-result.js -> c:/dev/electron/dist/main.electron.js
25176: _registerSourceMap: webpack:///(webpack)/hot/poll.js -> c:/dev/electron/dist/main.electron.js
25176: _registerSourceMap: webpack:///src-main/main.ts -> c:/dev/electron/dist/main.electron.js
```

**Comment by pushkin-** [comment]
@weinand Above, you said `<runtimeExecutable> <runtimeArgs> --debug-brk <program> <args>`.

Is that documented somewhere? It seems that no matter if I specify the arguments in `args` or `runtimeArgs` it puts the `--debug-brk` argument before them. Does this sound like a separate issue?

**Comment by weinand** [comment]
@pushkin- you can avoid the `--debug-brk` if you set the debug port explicitly, e.g. `"port": 5858`.
It is documented here: https://code.visualstudio.com/docs/editor/node-debugging#_launch-configuration-support-for-npm-and-other-tools

**Comment by pushkin-** [comment]
@weinand Let me clarify - I could have worded it a little better. The problem isn't that there's a `--debug-brk` flag, but that the runtime arguments are placed after that flag.

**Comment by weinand** [comment]
@pushkin- if there is a way to suppress the  `--debug-brk`, then it doesn't matter where arguments are placed, right? 
The command line without the `--debug-brk` is constructed like this:
```
<runtimeExecutable> <runtimeArgs> <program> <args>
```
So it doesn't matter how you use <runtimeArgs> <program> <args>.

**Comment by okoala** [comment]
https://github.com/eggjs/egg/issues/1052#issuecomment-308424839

**Comment by weinand** [comment]
This issue covers different issues and I have lost track of some of them. At least two issues have been resolved:
- it is no longer necessary to handle the "*.cmd" extension of Windows executables in a special way. They should be used automatically.
- in the Nov 2017 release runtime arguments are placed before the "--debug-brk" flag.
